### PR TITLE
defaultPosition now allows strings

### PR DESCRIPTION
**Pull request by tnrich**
This is a copy of https://github.com/mzabriskie/react-draggable/pull/271

but with the changes @STRML requested

### CODE REVIEW

**Comment by tnrich** [comment]
@STRML @eric-burel I fixed a couple issues in the commit I just made. Linting and tests now pass and I added a test for the new use case.

**Comment by tnrich** [comment]
@bokuweb when this PR gets merged this issue https://github.com/bokuweb/react-rnd/issues/437 will no longer be relevant. I should just be able to pass "50%" for both x and y defaultPositions

**Comment by tnrich** [comment]
@STRML any idea why this would be failing only on IE? 

That's the new spec I added: 
IE 11.0.0 (Windows 8.1 0.0.0) react-draggable props should render with defaultPosition set as string transform and handle subsequent translate() for DOM nodes FAILED
	AssertionError:   # specs\draggable.spec.jsx:255
	  
	  assert(style.indexOf('translate(calc(10% + 100px), calc(10% + 100px));') >= 0)
	         |     |                                                           |    
	         |     -1                                                          false
	         "touch-action: none;"                                                  
	  
	   at fail (specs/main.js:27336:3)
	   at ok (specs/main.js:27356:15)
	   at Decorator.prototype._callFunc (specs/main.js:30707:9)
	   at Decorator.prototype.concreteAssert (specs/main.js:30700:5)
	   at decoratedAssert (specs/main.js:31011:13)
	   at powerAssert (specs/main.js:4743:13)
	   at Anonymous function (specs/main.js:6344:7)

**Comment by tnrich** [comment]
hmm looks like it is related to this: 
image

I'm not sure if there is some way around this..?

**Comment by tnrich** [comment]
Okay found a solution based on this stackoverflow answer: https://stackoverflow.com/questions/21142923/ie-10-11-css-transitions-with-calc-do-not-work

this commit fixes it and tests should now pass :) 
https://github.com/mzabriskie/react-draggable/pull/361/commits/cf6979a81a3cd717dc464170fca0b1d3afbdf614


>updating createCSSTransform and createSVGTransform to use two translates instead of a calc function do to IE incompatibility; updating spec to reflect changes

**Comment by tnrich** [comment]
@mzabriskie @STRML any thoughts?

**Comment by tnrich** [comment]
bumping. @STRML is this project still active?

**Comment by tnrich** [comment]
@STRML bump

File: lib/utils/domFns.js [hunk 1]
```diff
@@ -109,13 +109,23 @@ export function offsetXYFromParent(evt: {clientX: number, clientY: number}, offs
   return {x, y};
 }
 
-export function createCSSTransform({x, y}: {x: number, y: number}): Object {
+export function createCSSTransform({x, y}: {x: number, y: number}, defaultPosition: {x: number | string, y: number | string }): Object {
```
Comment (STRML) [inline lib/utils/domFns.js]: You create a `DefaultControlPosition` type above but then don't use it here.

File: lib/utils/domFns.js [hunk 2]
```diff
@@ -109,13 +109,23 @@ export function offsetXYFromParent(evt: {clientX: number, clientY: number}, offs
...
+  const defaultX = (typeof defaultPosition.x === 'string') ? defaultPosition.x : defaultPosition.x + 'px';
+  const defaultY = (typeof defaultPosition.y === 'string') ? defaultPosition.y : defaultPosition.y + 'px';
+  const translation = (defaultPosition && defaultPosition.x !== 0 && defaultPosition.y !== 0)
+  ? `translate(${defaultX}, ${defaultY}) translate(${x}px,${y}px)`
+  : 'translate(' + x + 'px,' + y + 'px)';
   // Replace unitless items with px
```
Comment (STRML) [inline lib/utils/domFns.js]: This comment no longer makes sense with the changes above.

File: lib/utils/domFns.js [hunk 3]
```diff
@@ -109,13 +109,23 @@ export function offsetXYFromParent(evt: {clientX: number, clientY: number}, offs
   return {x, y};
 }
 
-export function createCSSTransform({x, y}: {x: number, y: number}): Object {
+export function createCSSTransform({x, y}: {x: number, y: number}, defaultPosition: {x: number | string, y: number | string }): Object {
+  const defaultX = (typeof defaultPosition.x === 'string') ? defaultPosition.x : defaultPosition.x + 'px';
```
Comment (STRML) [inline lib/utils/domFns.js]: This is really difficult to read. You create these two consts, `defaultX` and `defaultY`, but they're only actually used if and only if a `defaultPosition` is actually present. 

You then use that just to prepend what is being returned as a second translate. 

This is a lot of logic in what was previously a simple function and should be refactored such that the caller prepends the default translate if needed.

File: lib/utils/domFns.js [hunk 4]
```diff
@@ -109,13 +109,23 @@ export function offsetXYFromParent(evt: {clientX: number, clientY: number}, offs
...
+  const defaultY = (typeof defaultPosition.y === 'string') ? defaultPosition.y : defaultPosition.y + 'px';
+  const translation = (defaultPosition && defaultPosition.x !== 0 && defaultPosition.y !== 0)
+  ? `translate(${defaultX}, ${defaultY}) translate(${x}px,${y}px)`
+  : 'translate(' + x + 'px,' + y + 'px)';
```
Comment (STRML) [inline lib/utils/domFns.js]: Why no template string here?

File: lib/utils/domFns.js [hunk 5]
```diff
@@ -109,13 +109,23 @@ export function offsetXYFromParent(evt: {clientX: number, clientY: number}, offs
...
+  return {[browserPrefixToKey('transform', browserPrefix)]: translation };
 }
 
-export function createSVGTransform({x, y}: {x: number, y: number}): string {
-  return 'translate(' + x + ',' + y + ')';
+export function createSVGTransform({x, y}: {x: number, y: number}, defaultPosition: {x: number | string, y: number | string}): string {
+  const defaultX = (typeof defaultPosition.x === 'string') ? defaultPosition.x : defaultPosition.x + 'px';
+  const defaultY = (typeof defaultPosition.y === 'string') ? defaultPosition.y : defaultPosition.y + 'px';
+  const translation = (defaultPosition && defaultPosition.x !== 0 && defaultPosition.y !== 0)
+  ? `translate(${defaultX}, ${defaultY}) translate(${x}px,${y}px)`
```
Comment (STRML) [inline lib/utils/domFns.js]: This is wrong, svg transforms shouldn't have `px`.

**Review by STRML — CHANGES_REQUESTED** [review]

File: lib/utils/domFns.js (hunk 5 above)
Comment (tnrich) [reply lib/utils/domFns.js]: This should no longer have `px` in it

File: lib/utils/domFns.js (hunk 4 above)
Comment (tnrich) [reply lib/utils/domFns.js]: Added template strings everywhere I could

File: lib/utils/domFns.js (hunk 3 above)
Comment (tnrich) [reply lib/utils/domFns.js]: Reworked this to hopefully be more clear

File: lib/utils/domFns.js (hunk 2 above)
Comment (tnrich) [reply lib/utils/domFns.js]: Removed the comment

File: lib/utils/domFns.js (hunk 1 above)
Comment (tnrich) [reply lib/utils/domFns.js]: I now use it here

**Comment by tnrich** [comment]
@STRML let me know if there is anything else I can change to get this PR accepted and merged? Thanks!

**Comment by pavog** [comment]
Any updates on this? @tnrich @STRML

**Comment by tnrich** [comment]
@pavog I believe I made all the changes @STRML requested the first time around. We're just waiting on him?

**Comment by pavog** [comment]
@tnrich Yep, I believe that too. @STRML is still active (see #389) in this repository but seems to not want to review your changes again.

**Review by STRML — APPROVED** [review]

**Comment by pavog** [comment]
@STRML Thank you very much 👍

**Comment by eric-burel** [comment]
Awesome, thanks all for finishing this PR and merging !

**Comment by tnrich** [comment]
@eric-burel @STRML had to revert the changes due to unexpected breakages. I've made a new PR that hopefully solves those issues by introducing a new property `initialPosition`  

https://github.com/mzabriskie/react-draggable/pull/393

**Pull request merged**